traffic_ops/app/db/reencrypt/reencrypt.go

```diff
@@ -0,0 +1,393 @@
+		die("previous-key flag is required.")
+	}
+	if newKeyLocation == nil || *newKeyLocation == "" {
+		die("new-key flag is required.")
+	}
+
+	newKey, err := readKey(*newKeyLocation)
+	if err != nil {
+		die("reading new-key: " + err.Error())
+	}
+
+	previousKey, err := readKey(*previousKeyLocation)
+	if err != nil {
+		die("reading previous-key: " + err.Error())
+	}
+
+	dbConfBytes, err := ioutil.ReadFile(*cfg)
+	if err != nil {
+		die("reading db conf '" + *cfg + "': " + err.Error())
+	}
+
+	pgCfg := Config{}
+	err = json.Unmarshal(dbConfBytes, &pgCfg)
+	if err != nil {
+		die("unmarshalling '" + *cfg + "': " + err.Error())
+	}
+
+	sslStr := "require"
+	if !pgCfg.SSL {
+		sslStr = "disable"
+	}
+	db, err := sqlx.Open("postgres", fmt.Sprintf("postgres://%s:%s@%s:%d/%s?sslmode=%s&fallback_application_name=trafficvault", pgCfg.User, pgCfg.Password, pgCfg.Hostname, pgCfg.Port, pgCfg.DBName, sslStr))
+	if err != nil {
+		die("opening database: " + err.Error())
+	}
+
+	if err = reEncryptSslKeys(db, previousKey, newKey); err != nil {
+		die("re-encrypting SSL Keys: " + err.Error())
+	}
+	if err = reEncryptUrlSigKeys(db, previousKey, newKey); err != nil {
+		die("re-encrypting URL Sig Keys: " + err.Error())
+	}
+	if err = reEncryptUriSigningKeys(db, previousKey, newKey); err != nil {
+		die("re-encrypting URI Signing Keys: " + err.Error())
+	}
+	if err = reEncryptDNSSECKeys(db, previousKey, newKey); err != nil {
+		die("re-encrypting DNSSEC Keys: " + err.Error())
+	}
+
+	if err = updateKeyFile(previousKey, *previousKeyLocation, newKey); err != nil {
+		die("updating the key file: " + err.Error())
+	}
+
+	fmt.Println("Successfully re-encrypted keys for SSL Keys, URL Sig Keys, URI Signing Keys, and DNSSEC Keys.")
+}
+
+type Config struct {
+	DBName   string `json:"dbname"`
+	Hostname string `json:"hostname"`
+	User     string `json:"user"`
+	Password string `json:"password"`
+	Port     int    `json:"port"`
+	SSL      bool   `json:"ssl"`
+}
+
+func readKey(keyLocation string) ([]byte, error) {
+	var keyBase64 string
+	keyBase64Bytes, err := ioutil.ReadFile(keyLocation)
+	if err != nil {
+		return []byte{}, errors.New("reading file '" + keyLocation + "':" + err.Error())
+	}
+	keyBase64 = string(keyBase64Bytes)
+
+	key, err := base64.StdEncoding.DecodeString(keyBase64)
+	if err != nil {
+		return []byte{}, errors.New(fmt.Sprintf("AES key cannot be decoded from base64: %v", err))
+	}
+
+	// verify the key works
+	_, err = aes.NewCipher(key)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return key, nil
+}
+
+func reEncryptSslKeys(db *sqlx.DB, previousKey []byte, newKey []byte) error {
+	tx, err := db.Begin()
+	if err != nil {
+		return errors.New(fmt.Sprintf("transaction begin failed %v %v ", err, tx))
+	}
+	defer tx.Commit()
+
+	rows, err := tx.Query("SELECT id, data FROM sslkey")
+	if err != nil {
+		return errors.New(fmt.Sprintf("querying: %v", err))
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		updateTx, err := db.Begin()
```

Comment:
Rather than creating two transactions, you can probably just read all of the rows into map. Then once all the rows have been read into the map, you should be able to use the same transaction to do the updates.

This comment applies to the other functions as well.